### PR TITLE
Add a initial_best parameter in ModelCheckpoint Callback

### PR DESCRIPTION
**Pull request by old-school-kid**
Add a initial_best parameter in ModelCheckpoint that is stored as self.best. 

**Problem** If a session is stopped and rerun or kernel dies, then the best value of metric is taken as np.inf or -np.inf and the model is saved after the first epoch without taking into consideration the initial performance of model. So modelweights are overwritten even if the previous model performance was better than the current model.

**Solution** Instead of assigning inf or -inf to self.best the initial_best parameter is saved as self.best so that it doesnt overwrite the model.

Unit test can be found in [here](https://colab.research.google.com/drive/1MpennD9t136peziuCjef7TwbVendMZcS?usp=sharing)

### CODE REVIEW

**Comment by gbaned** [comment]
@old-school-kid Can you please check @deeb02's comments and keep us posted ? Thanks!

**Comment by old-school-kid** [comment]
I am sorry for the delay as I had some academic commitments. 

The general use of the feature has been explained in this [notebook](https://colab.research.google.com/drive/1MpennD9t136peziuCjef7TwbVendMZcS?usp=sharing)

Unit test for the feature can be found again in this [notebook](https://colab.research.google.com/drive/1RWATYkBUnNaJhPcULtN7jxmB8YUXVrj1?usp=sharing)
Hooks for default setting and then with montor being accuracy and loss with mode being auto, min and max have been run. 

@deeb02 can you please give it a look?
Thanks

**Comment by fchollet** [comment]
@old-school-kid independently of whether you have tests in a separate notebook, please add unit tests in the PR. The main purpose of unit tests is not to check that a feature works correctly when it gets initially added, but to automatically check that it keeps working correctly in the future.

**Comment by old-school-kid** [comment]
Test for the following cases has been added.
mode = 'max'
monitor = 'val_acc'
initial_best = 0
expected behaviour: Saves model

mode = 'auto'
monitor = 'val_loss'
initial_best = None
expected behaviour: Saves model

mode = 'min'
monitor = 'val_loss'
initial_best = 0
expected behaviour: Doesnt save model

mode = 'auto'
monitor = 'val_loss'
initial_best = 0
expected behaviour: Doesnt save model

In my knowledge this would a good sample from all the possible combinations. A detailed test with all possible combinations can be found in [this notebook](https://colab.research.google.com/drive/1RWATYkBUnNaJhPcULtN7jxmB8YUXVrj1?usp=sharing)

File: keras/callbacks.py [hunk 1]
```diff
@@ -1253,6 +1253,9 @@ class ModelCheckpoint(Callback):
       options: Optional `tf.train.CheckpointOptions` object if
         `save_weights_only` is true or optional `tf.saved_model.SaveOptions`
         object if `save_weights_only` is false.
+      initial_best: Initial "best" value of the metric to be monitored.
```
Comment (fchollet) [inline keras/callbacks.py:1256]: Let's rename this "initial_value_threshold" -- thanks!

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for adding the tests. I pushed some copyedits. Please pull them first.

**Comment by old-school-kid** [comment]
@fchollet 
Yeah sounds good. Did the necessary changes.

File: keras/callbacks.py (hunk 1 above)
Comment (fchollet) [reply keras/callbacks.py:1256]: Please update the docstrings as well.

Comment (old-school-kid) [reply keras/callbacks.py]: @fchollet made the changes in docstring.

File: keras/callbacks.py [hunk 2]
```diff
@@ -1278,6 +1282,7 @@ def __init__(self,
     self.epochs_since_last_save = 0
     self._batches_seen_since_last_saving = 0
     self._last_batch_seen = 0
+    self.best= initial_value_threshold
```
Comment (mattdangerw) [inline keras/callbacks.py:1285]: rename best -> initial_value_threshold everywhere, and add a space before the = sign

Comment (mattdangerw) [reply keras/callbacks.py:1285]: Oh whoops, I see this was named `best` before, and is updated elsewhere. Name is fine! But please do fix the formatting mistake.

Comment (old-school-kid) [reply keras/callbacks.py]: @mattdangerw 
Thank you for pointing out the mistake. Can you take a look now?

**Review by mattdangerw — APPROVED** [review]

**Comment by old-school-kid** [comment]
@gbaned 
All the requested changes have been made and yet it says _changes requested_. Can you please let me know what else has to be done to expedite the process?
Thank you

File: keras/callbacks.py [hunk 3]
```diff
@@ -1253,6 +1253,9 @@ class ModelCheckpoint(Callback):
       options: Optional `tf.train.CheckpointOptions` object if
         `save_weights_only` is true or optional `tf.saved_model.SaveOptions`
         object if `save_weights_only` is false.
+      initial_value_threshold: Initial "best" value of the metric to be monitored.
```
Comment (mattdangerw) [inline keras/callbacks.py:1256]: this line is over 80 characters.

File: keras/callbacks.py [hunk 4]
```diff
@@ -1253,6 +1253,9 @@ class ModelCheckpoint(Callback):
...
+        Only overwrites the model weights already saved if the
+        performance of current model is better than this value.
```
Comment (mattdangerw) [inline keras/callbacks.py:1258]: This only applies when save_best_only is true correct? We should call that out. Something like (add newlines).

Initial "best" value of the metric to be monitored. Only applies if `save_best_value=True`. If set, the checkpoint will only be saved if the model metric value is better than this value.

File: keras/callbacks_test.py [hunk 5]
```diff
@@ -848,6 +848,101 @@ def test_ModelCheckpoint(self):
...
+    mode = 'max'
+    monitor = 'val_acc'
+    initial_value_threshold = 0
+    cbks = [
+        keras.callbacks.ModelCheckpoint(
+            filepath,
+            monitor=monitor,
+            save_best_only=save_best_only,
+            initial_value_threshold=initial_value_threshold,
+            save_best_only=True,
```
Comment (mattdangerw) [inline keras/callbacks_test.py:862]: save_best_only is set twice (here and elsewhere).

**Review by mattdangerw — CHANGES_REQUESTED** [review]
Found some linting errors and other mistakes while trying to merge this!

File: keras/callbacks.py (hunk 4 above)
Comment (old-school-kid) [reply keras/callbacks.py]: Thank you for pointing out that. Would you please check it again?

**Review by mattdangerw — APPROVED** [review]
Thanks! Will try landing again.

File: keras/callbacks.py (hunk 4 above)
Comment (old-school-kid) [reply keras/callbacks.py]: Thank you for pointing that out. Could you please take a look again?

**Comment by old-school-kid** [comment]
@mattdangerw is it breaking somewhere again?

**Comment by mattdangerw** [comment]
A few more linting errors (line length) but I will fix those up while landing. As this is an API change we need to generate some API files and gather approvals, but no action needed here I hope. This will auto close when the PR lands.

**Comment by old-school-kid** [comment]
Thank you @mattdangerw

**Review by mattdangerw — CHANGES_REQUESTED** [review]
Actually it looks like these new tests are broken. We had a misconfiguration on github so these tests were actually being exercised, but if you resync this should be fixed. I don't believe the tests added here ever passed.

Here's a stack trace. Can you take a look?

```
Traceback (most recent call last):
  File ".../keras/callbacks.py", line 1479, in _get_file_path
    file_path = self.filepath.format(epoch=epoch + 1, **logs)
KeyError: 'batch'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File ".../absl/testing/parameterized.py", line 316, in bound_param_test
    return test_method(self, *testcase_params)
  File ".../keras/keras_parameterized.py", line 282, in decorated
    _test_functional_model_type(f, self, *args, **kwargs)
  File ".../keras/keras_parameterized.py", line 296, in _test_functional_model_type
    f(test_or_class, *args, **kwargs)
  File ".../keras/callbacks_test.py", line 872, in test_ModelCheckpoint
    verbose=0)
  File ".../keras/utils/traceback_utils.py", line 60, in error_handler
    return fn(*args, **kwargs)
  File ".../keras/engine/training.py", line 1360, in fit
    callbacks.on_epoch_end(epoch, epoch_logs)
  File ".../keras/callbacks.py", line 414, in on_epoch_end
    callback.on_epoch_end(epoch, logs)
  File ".../keras/callbacks.py", line 1386, in on_epoch_end
    self._save_model(epoch=epoch, batch=None, logs=logs)
  File ".../keras/callbacks.py", line 1421, in _save_model
    filepath = self._get_file_path(epoch, batch, logs)
  File ".../keras/callbacks.py", line 1485, in _get_file_path
    f'Failed to format this callback filepath: "{self.filepath}". '
KeyError: 'Failed to format this callback filepath: ".../tmp/tmp4k47f9qg/checkpoint.epoch{epoch:02d}batch{batch:02d}.h5". Reason: \'batch\''
```

**Comment by old-school-kid** [comment]
Hi @mattdangerw 
filepath was initialized in line number 543 in the test. But it is again formatted in line 693 to accomodate epoch number in the name which raises error.  Initialized it again with any formatting. Would you please try again?
Thank you

**Comment by mattdangerw** [comment]
Thanks! That did it.

**Comment by old-school-kid** [comment]
@mattdangerw 
The same issue this PR solves can be found in [EarlyStopping Callback](https://github.com/keras-team/keras/blob/5d2993b78603bda911f0a4912d08b1e5f36eb633/keras/callbacks.py#L1822)
If the instance is killed then early stopping takes the best value as inf and continues training even if the first epoch performs worse than the last saved model and if `restore_best_weights=True` it overwrites the model.
So should I submit another PR to solve this issue?

**Comment by mattdangerw** [comment]
@old-school-kid I think that might need more thought. You can't really assume early stopping is being called on a previously saved model right? And the model will continue to update its weights as its going, and will need something to use for the set_weights call... What would you do if a model is trained from scratch and no model beats the threshold?

I would open an issue to discuss, list the new docstrings and explain how the argument would work. For now I will close this, as this has been landed. Thanks for PR!

**Pull request closed without merging**